src/digraph/Dominator_Tree.jl

```diff
@@ -0,0 +1,181 @@
+  if there is not such a node we will consider the parent of the node of interst to be the semi dominator of it
+   there is a duality between the semi dominator and the  immediate dominator 
+   min(preorderof(semi_dominator(v)))>=max(preorder(dominator(v)) 
+   for more details see the paper
+   the key theorm is 
+   Let w != the source and let u be a vertex for which semi_dominator(u) is minimum 
+   among vertices between w and its semi dominator. 
+   Then the immediate dominator is  
+   semi_dominator(w) if semi_dominator(w) = semi_dominator(u), 
+   immidiate_dominator(w) =  immidiate_dominator(u) otherwise.
+  =#
+  semi=zeros(T,nv(g))
+  #==# 
+   
+  #=
+   we will store each node in the bucket of its semi dominator,
+   untill we we figure out all the semi dominator of the nodes between them 
+  =#       
+     
+  buctes=Array{Array,1}(undef,nv(g)) 
+
+  verticesar=zeros(T,nv(g)) #map between each preorder of a node and its real name 
+  dom=zeros(T,nv(g))        #the immediate dominator of the the node 
+  #=
+  one important step in the algorithm is to know a  vertex u for which semi_dominator(u) is minimum 
+  among vertices between every node and its semi dominator,
+  after we order the vertices by preorder, we will scan them backwards,
+  then we can imagine the graph as a forest of trees, in each step we merge two trees by making the root of one tree the 
+  child of the root of the other tree, now we decoverd another child of the later root,let it be w
+  then we will see if there is node u that consider w to be its semi dominator, and look for the smalest smei dominator in the path between w and u
+  we can leverage some facts
+  - we donot need to travel all the path each time, we can do path decompsion
+  - each time we do a merge one of the two trees will not be affected at all, 
+  in the other tree if there is a node that has semi dominator smaller than the new root itse descendants will not be affected
+      
+  =#
+  
+  childs=zeros(T,nv(g)) # the  child u of a node w of the the node that its suntree has a big size and its has a semi_dominator snaller than w 
+  ancesstor=zeros(T,nv(g)) # the  ancesstor u of node w is the cuurent parent of w in the forest  
+  labels=zeros(T,nv(g))       # the label u of node w is the result of accumalteing some of its parents nodes in the operation of path decompostion 
+
+  parent[source]=1
+  #find the the node u the has the smalest semi_dominator in the path between v and the current root of its subtree in the current forest
+  function eval(v)
+
+    s=v
+    parents=ancesstor[s]
+    vstack=Vector{Int32}([])
+
+    if parents==0
+      return labels[s]
+    end
+    while ancesstor[parents]!=0
+          push!(vstack,s)
+          s=parents
+          parents=ancesstor[parents]
+        end
+        boss=parents
+        parents=s
+        len=length(vstack)
+        while len!=0
+          len-=1
+          s=pop!(vstack)
```

Comment:
since you're not using `len` except to check to see whether `vstack` is empty, why not just `while !isempty(vstack)` ?